### PR TITLE
VDR: Support root did:web DIDs

### PR DESCRIPTION
**Pull request by reinkrul**
- Renames API's `GetWebDID` to `GetUserWebDID`
- Adds `GetRootWebDID` that uses the `Host` header to determine the requested DID
- Changes in `CreateDID` API:
  - Renames the `id` parameter to `user_id` to more accurately describe the concept (ID of a DID is the part that follows `did:web:`)
  - Adds the `did` parameter which can be used to specify the full DID to be created. This can either be a rooted did:web, or with user path. If the user path is specified, it must follow the pattern `did:web:<host>:iam:<user ID>`.

Requires https://github.com/nuts-foundation/nuts-node/issues/2903 to make the OAuth2/OpenID endpoints available for root did:web DIDs.

Fixes https://github.com/nuts-foundation/nuts-node/issues/2833

TODO:
- [x] Documentation
- [x] e2e test with OAuth2/OpenID flow

### CODE REVIEW

File: auth/api/iam/api.go [hunk 1]
```diff
@@ -595,14 +626,27 @@ func (r Wrapper) StatusList(ctx context.Context, request StatusListRequestObject
...
 }
 
-// idToDID converts the tenant-specific part of a did:web DID (e.g. 123)
-// to a fully qualified did:web DID (e.g. did:web:example.com:123), using the configured Nuts node URL.
-func (r Wrapper) idToDID(id string) did.DID {
-	identityURL := r.auth.PublicURL().JoinPath("iam", id)
+// requestedDID constructs a did:web DID as it was requested by the API caller. It can be a DID with or without user path, e.g.:
+// - did:web:example.com
+// - did:web:example:iam:1234
+// When host is given, it is used as base for the did:web DID. If it's absent, the configured URL is used as base.
```
Comment (reinkrul) [inline auth/api/iam/api.go:632]:
```suggestion
```

File: auth/api/iam/api.go [hunk 2]
```diff
@@ -400,7 +400,24 @@ func (r *Wrapper) validateJARRequest(ctx context.Context, rawToken string, clien
...
+	didAsString := r.requestedDID(request.Id).String()
+	md, err := r.oauthAuthorizationServerMetadata(ctx, didAsString)
+	if err != nil {
+		return nil, err
+	}
+	return OAuthAuthorizationServerMetadata200JSONResponse(*md), nil
+}
+
+func (r Wrapper) RootOAuthAuthorizationServerMetadata(ctx context.Context, request RootOAuthAuthorizationServerMetadataRequestObject) (RootOAuthAuthorizationServerMetadataResponseObject, error) {
+	md, err := r.oauthAuthorizationServerMetadata(ctx, r.requestedDID("").String())
```
Comment (woutslakhorst) [inline auth/api/iam/api.go:412]:
```suggestion
	md, err := r.oauthAuthorizationServerMetadata(ctx, r.auth.PublicURL().String())
```

File: docs/_static/vdr/v2.yaml [hunk 3]
```diff
@@ -367,11 +367,27 @@ components:
...
-          id:
+          did:
             type: string
-            description: The ID of the DID document. If not given, a random UUID is generated.
-            example: "did:web:example.com:iam:013c6fda-73e8-45ee-9220-48652dba854b"
+            description: |
+              The DID of the DID document. If it's a did:web DID and it contains a user path, it must follow the format `did:web:example.com:iam:1234`.
+              It can be used, for instance, to create a root did:web DID (one without user path).
+            example:
+              - "did:web:example.com"
```
Comment (woutslakhorst) [inline docs/_static/vdr/v2.yaml:380]: How does this work with only one publicURL configured?

File: auth/api/iam/api.go [hunk 4]
```diff
@@ -413,20 +430,34 @@ func (r Wrapper) OAuthAuthorizationServerMetadata(ctx context.Context, request O
...
-		log.Logger().WithError(err).Errorf("Could not resolve Web DID: %s", ownDID.String())
+		log.Logger().WithError(err).Errorf("Could not resolve user did:web: %s", ownDID.String())
 		return nil, errors.New("unable to resolve DID")
 	}
-	return GetWebDID200JSONResponse(*document), nil
+	return GetUserWebDID200JSONResponse(*document), nil
+}
+
+func (r Wrapper) GetRootWebDID(ctx context.Context, _ GetRootWebDIDRequestObject) (GetRootWebDIDResponseObject, error) {
+	ownDID := r.requestedDID("")
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply auth/api/iam/api.go:451]: I still think we might do away with `url`, so this is then the right abstraction? Also, it then works the same way as with a tenant

File: vdr/didweb/manager.go [hunk 5]
```diff
@@ -32,21 +32,30 @@ import (
...
+	"strings"
 )
 
 func DefaultCreationOptions() management.CreationOptions {
 	return management.Create(MethodName)
 }
 
-type userPathOption struct {
-	path string
+type userIDOption struct {
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply vdr/didweb/manager.go:42]: renamed to tenant

File: docs/_static/vdr/v2.yaml (hunk 3 above)
Comment (reinkrul) [reply docs/_static/vdr/v2.yaml:380]: The easy/most flexible API is: "I want to create this DID: <DID>"

I think we might support multiple domains in future, then this API still applies. I'd add a check that verified the given DID follows the configured `url`.

**Review by woutslakhorst — COMMENTED** [review]
missing todo

is it possible to create `example.com/iamnot/5`?

**Comment by reinkrul** [comment]
> is it possible to create `example.com/iamnot/5`?

No, see `vdr/didweb/manager.go:133`, `create()` checks the DID to create conforms to the configured `url` and that it follows `<iam>:<tenant>`, if it contains a path.

File: auth/api/iam/api.go (hunk 2 above)
Comment (reinkrul) [reply auth/api/iam/api.go:412]: requires a DID

**Comment by reinkrul** [comment]
Documentation-wise is the OpenAPI specification enough for now. If there's specific situations in which root did:web is especially important, it can be documented on the Wiki

**Review by woutslakhorst — APPROVED** [review]

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**